File: ts/src/production-traces/taxonomy/anthropic-error-reasons.ts
/**
 * Anthropic exception class → `outcome.error.type` taxonomy (TS half).
 *
 * Cross-runtime parity: the Python counterpart at
 * `autocontext/src/autocontext/production_traces/taxonomy/anthropic_error_reasons.py`
 * MUST have the same keys + values. Parity tests keep the two in lock-step.
 */

export type AnthropicErrorReasonKey =
  | "rateLimited"
  | "timeout"
  | "badRequest"
  | "authentication"
  | "permissionDenied"
  | "notFound"
  | "apiConnection"
  | "overloaded"
  | "upstreamError"
  | "uncategorized";

export const ANTHROPIC_ERROR_REASONS: Readonly<
  Record<string, AnthropicErrorReasonKey>
> = Object.freeze({
  RateLimitError: "rateLimited",
  APITimeoutError: "timeout",
  BadRequestError: "badRequest",
  AuthenticationError: "authentication",
  PermissionDeniedError: "permissionDenied",
  NotFoundError: "notFound",
  APIConnectionError: "apiConnection",
  OverloadedError: "overloaded",
  ConflictError: "upstreamError",
  UnprocessableEntityError: "upstreamError",
  InternalServerError: "upstreamError",
  APIStatusError: "upstreamError",
  APIError: "upstreamError",
});

export const ANTHROPIC_ERROR_REASON_KEYS: readonly AnthropicErrorReasonKey[] =
  Object.freeze([
    "rateLimited",
    "timeout",
    "badRequest",
    "authentication",
    "permissionDenied",
    "notFound",
    "apiConnection",
    "overloaded",
    "upstreamError",
    "uncategorized",
  ]);

File: autocontext/src/autocontext/production_traces/taxonomy/anthropic_error_reasons.py
"""Anthropic exception class → ``outcome.error.type`` taxonomy.

Parity with TS at ``ts/src/production-traces/taxonomy/anthropic-error-reasons.ts``.
Class names stored as strings so the table imports cleanly across SDK versions —
``OverloadedError`` exists throughout Anthropic 0.x but we shouldn't bind to it
at import time.
"""
from __future__ import annotations

from types import MappingProxyType
from typing import Final, Literal

AnthropicErrorReasonKey = Literal[
    "rateLimited",
    "timeout",
    "badRequest",
    "authentication",
    "permissionDenied",
    "notFound",
    "apiConnection",
    "overloaded",
    "upstreamError",
    "uncategorized",
]

_RAW: Final = {
    "RateLimitError": "rateLimited",
    "APITimeoutError": "timeout",
    "BadRequestError": "badRequest",
    "AuthenticationError": "authentication",
    "PermissionDeniedError": "permissionDenied",
    "NotFoundError": "notFound",
    "APIConnectionError": "apiConnection",
    "OverloadedError": "overloaded",
    "ConflictError": "upstreamError",
    "UnprocessableEntityError": "upstreamError",
    "InternalServerError": "upstreamError",
    "APIStatusError": "upstreamError",
    "APIError": "upstreamError",
}

ANTHROPIC_ERROR_REASONS: Final = MappingProxyType(_RAW)
ANTHROPIC_ERROR_REASON_KEYS: Final = frozenset(_RAW.values()) | {"uncategorized"}
